### PR TITLE
Refactor defense mode

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

 ```SUMMARY: Infrastructure "Refactor defense mode"```

#### Purpose of change

Defense mode needs some love. I've made some internal code refactor to simplify future changes.

#### Describe the solution

See commit messages.

#### Testing

Start game in defense mode, see updated menu.

#### Additional context

Future updates will follow eventually.

#### Screenshots

** Defense mode setup menu (before) **
image

** Defense mode setup menu (after) **

image

### CODE REVIEW

File: src/gamemode_defense.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        // Do we need to fulfill thirst?
+        bool thirst;
+        // Do we need to sleep?
+        bool sleep;
+
+        // Do caravans offer the option of hiring a mercenary?
+        bool mercenaries;
+
+        // Allow save
+        bool allow_save =  true;
```
Comment (Night-Pryanik) [inline src/gamemode_defense.h]:
```suggestion
        bool allow_save = true;
```

Also, if I'm not mistaken, `allow_save` is only set once to `true` here, and never changes anywhere?

File: src/gamemode_defense.cpp [hunk 2]
```diff
@@ -143,43 +144,78 @@ void defense_game::per_turn()
...
+        case ACTION_SLEEP:
+            if( !sleep ) {
+                action_error_message = _( "You don't need to sleep!" );
+            }
+            break;
+        case ACTION_SAVE:
+        case ACTION_QUICKSAVE:
+            if( !allow_save ) {
+                add_msg( m_info, _( "You cannot save in defense mode!" ) );
+                act = ACTION_NULL;
```
Comment (BevapDin) [inline src/gamemode_defense.cpp]: This seems redundant. The other cases don't have it either.

File: src/overmap.cpp [hunk 3]
```diff
@@ -434,6 +434,16 @@ const std::vector<overmap_special> &overmap_specials::get_all()
     return specials.get_all();
 }
 
+const overmap_special overmap_specials::get_specific( const std::string &id )
```
Comment (BevapDin) [inline src/overmap.cpp]:
```suggestion
const overmap_special overmap_specials::get_specific( const overmap_special_id &id )
```
And get rid of the call to `str()` in the loop below. `id` is an identifier referring to specific kind of thing: not a monster type, not a terrain type, but an overmap_special type.

----

The function isn't actually needed. You should use the established way to get the object that an id refers to: call `string_id::obj` or `string_id::operator*` - that should work for all types of ids:
```C++
const overmap_special_id id("foo");
const overmap_special &sp = *id;
```

File: src/gamemode_defense.h (hunk 1 above)
Comment (ZhilkinSerg) [reply src/gamemode_defense.h]: I forgot to expose option to UI.

File: src/overmap.cpp (hunk 3 above)
Comment (ZhilkinSerg) [reply src/overmap.cpp]: Thanks. That's much more cleaner.

File: src/gamemode_defense.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        bool mercenaries;
+
+        // Allow save
+        bool allow_save;
+
+        // Start defence location position on overmap
+        tripoint defloc_pos;
+
+        // Defense location special
+        overmap_special defloc_special;
```
Comment (BevapDin) [inline src/gamemode_defense.h]: You don't need to store a whole *copy* of the special location, the id itself is enough.
```suggestion
        overmap_special_id defloc_special;
```
You can get the object from the id whenever you need it (which seems to be just *one* place).

File: src/gamemode_defense.cpp [hunk 5]
```diff
@@ -317,7 +312,7 @@ void defense_game::init_map()
         }
     }
 
-    g->load_map( tripoint( 100, 100, 0 ) );
+    g->load_map( defloc_pos * 2 );
```
Comment (ZhilkinSerg) [inline src/gamemode_defense.cpp]:
```suggestion
    g->load_map( omt_to_sm_copy( defloc_pos ) );
```

File: src/init.cpp [hunk 6]
```diff
@@ -71,7 +71,6 @@
 #include "string_formatter.h"
 #include "text_snippets.h"
 #include "trap.h"
-#include "tutorial.h"
```
Comment (ZhilkinSerg) [inline src/init.cpp]:
```suggestion
    #include "gamemode_tutorial.h"
```

File: src/init.cpp [hunk 7]
```diff
@@ -71,7 +71,6 @@
...
 #include "veh_type.h"
```
Comment (ZhilkinSerg) [inline src/init.cpp]:
```suggestion
#include "gamemode_tutorial.h"
#include "veh_type.h"
```

File: src/gamemode_defense.cpp [hunk 8]
```diff
@@ -143,42 +146,74 @@ void defense_game::per_turn()
...
+        case ACTION_MOVE_E:
+            if( g->u.posx() == HALF_MAPSIZE_X + SEEX - 1 && g->get_levx() >= 98 ) {
+                leaving_defloc = true;
+            }
+            break;
+        case ACTION_MOVE_SE:
+            if( ( g->u.posy() == HALF_MAPSIZE_Y + SEEY - 1 && g->get_levy() >= 98 ) ||
+                ( g->u.posx() == HALF_MAPSIZE_X + SEEX - 1 && g->get_levx() >= 98 ) ) {
+                leaving_defloc = true;
+                break;
```
Comment (ZhilkinSerg) [inline src/gamemode_defense.cpp]:
```suggestion
```

File: src/gamemode_defense.cpp [hunk 9]
```diff
@@ -143,42 +146,74 @@ void defense_game::per_turn()
...
+            }
```
Comment (ZhilkinSerg) [inline src/gamemode_defense.cpp]:
```suggestion
            }
            break;
```

**Pull request merged**